rust-symcrypt/src/cipher/cbc.rs

```diff
@@ -0,0 +1,330 @@
+    pub fn aes_cbc_decrypt(
+        &self,
+        chaining_value: &mut [u8; AES_BLOCK_SIZE as usize],
+        cipher_text: &[u8],
+        plain_text: &mut [u8],
+    ) -> Result<(), SymCryptError> {
+        validate_block_size(plain_text, cipher_text)?;
+        unsafe {
+            symcrypt_sys::SymCryptAesCbcDecrypt(
+                self.expanded_key.get_inner(),
+                chaining_value.as_mut_ptr(),
+                cipher_text.as_ptr(),
+                plain_text.as_mut_ptr(),
+                plain_text.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+        Ok(())
+    }
+
+    /// `aes_cbc_decrypt_in_place()` decrypts the `buffer` using the AES-CBC algorithm and writes the decrypted data back to the `buffer` provided.
+    /// You are able to run this function in one go or block by block.
+    ///
+    /// `chaining_value` is a `mut &[u8; 16]` that represents the `IV` on the first call, and will be filled with the chaining value after the decryption is complete.
+    /// If doing decryption block by block, you will need to keep track of the `chaining_value` and pass it in on the next block for decryption. This value must be 16 bytes long.
+    ///
+    /// `buffer` is a `mut &[u8]` that will be filled with the decrypted data. The length of this buffer must be a multiple of 16 bytes.
+    ///
+    /// This function will return a [`SymCryptError`] if the length of the `buffer` is not a multiple of 16 bytes.
+    pub fn aes_cbc_decrypt_in_place(
+        &self,
+        chaining_value: &mut [u8; AES_BLOCK_SIZE as usize],
+        buffer: &mut [u8],
+    ) -> Result<(), SymCryptError> {
+        validate_block_size(buffer, buffer)?;
+        unsafe {
+            symcrypt_sys::SymCryptAesCbcDecrypt(
+                self.expanded_key.get_inner(),
+                chaining_value.as_mut_ptr(),
+                buffer.as_ptr(),
+                buffer.as_mut_ptr(),
+                buffer.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+pub mod test {
+    use super::*;
+    use crate::cipher::AES_BLOCK_SIZE;
+    use hex;
+    use std::convert::TryInto;
+
+    #[test]
+    fn test_aes_cbc_encrypt() {
+        let key = hex::decode("00000000000000000000000000000000").unwrap();
+        let aes_cbc = AesExpandedKey::new(&key).unwrap();
+        let mut chaining_value: [u8; 16] = hex::decode("00000000000000000000000000000000")
+            .unwrap()
+            .try_into()
+            .unwrap();
+        let plain_text = hex::decode("f34481ec3cc627bacd5dc3fb08f273e6").unwrap();
+        let mut cipher_text = vec![0u8; plain_text.len()];
+
+        aes_cbc
+            .aes_cbc_encrypt(&mut chaining_value, &plain_text, &mut cipher_text)
+            .unwrap();
+        assert_eq!(hex::encode(cipher_text), "0336763e966d92595a567cc9ce537f5e");
+    }
+
+    #[test]
+    fn test_aes_cbc_encrypt_single_and_piecewise() {
```

Comment:
Would probably be good to also have a piecewise / block-wise decryption test

Reply:
Added